### PR TITLE
gh-116122: Add SBOM generation to PCbuild/build.bat

### LINKED ISSUE #116122: Add SBOM regeneration to PCBuild/build.bat
Allows for Windows contributors to regenerate the SBOM file, common with the `cpython-source-deps` SBOM.

### Linked PRs
* gh-116138
* gh-118435

### PR DESCRIPTION
**Pull request by sethmlarson**
I manually wrote this as I don't have access to Windows, will need someone with access to that platform to test this configuration out.

Closes #116122 

* Issue: gh-116122

### CODE REVIEW

**Comment by zooba** [comment]
The build script side seems to be okay, but the SBOM script itself doesn't work. It looks like it's because git produces harder to parse output for this command:

```
D:\cpython> git check-ignore --verbose --non-matching ./Modules/_decimal/libmpdec\vcdiv64.asm
::      "./Modules/_decimal/libmpdec\\vcdiv64.asm"
```

Adding an `ast.literal_eval` call into `filter_gitignored_paths()` helped, and then adding `.replace(b'\r\n', b'\n')` to the checksum calculations fixed the line ending differences. A quick hack of `"fileName": str(PurePosixPath(PureWindowsPath(path)))` fixes the path formatting, and there'll be a better way to make that cross platform.

Once those are done, there are still a few files showing up for me that aren't already in the SBOM. I can't immediately see why - maybe they're just lingering in my checkout? Or possibly they're being omitted from the POSIX run because of case sensitivity?

* Modules/expat/expat_config.h
* Modules/_hacl/README.md
* Modules/_hacl/refresh.sh
* Lib/ctypes/macholib/README.ctypes
* Lib/ctypes/macholib/fetch_macholib
* Lib/ctypes/macholib/fetch_macholib.bat

File: PCbuild/regen.targets [hunk 1]
```diff
@@ -33,11 +33,15 @@
     <_JITOutputs Include="$(GeneratedPyConfigDir)jit_stencils.h"/>
     <_CasesSources Include="$(PySourcePath)Python\bytecodes.c;$(PySourcePath)Python\optimizer_bytecodes.c;"/>
     <_CasesOutputs Include="$(PySourcePath)Python\generated_cases.c.h;$(PySourcePath)Include\opcode_ids.h;$(PySourcePath)Include\internal\pycore_uop_ids.h;$(PySourcePath)Python\opcode_targets.h;$(PySourcePath)Include\internal\pycore_opcode_metadata.h;$(PySourcePath)Include\internal\pycore_uop_metadata.h;$(PySourcePath)Python\optimizer_cases.c.h;$(PySourcePath)Lib\_opcode_metadata.py"/>
+    <_SbomSources Include="$(PySourcePath)PCBuild\get_externals.bat" />
```
Comment (zooba) [inline PCbuild/regen.targets]:
```suggestion
    <_SbomSources Include="$(PySourcePath)PCbuild\get_externals.bat" />
```

Incredibly minor, but we've had people run into case issues before, so may as well make it match.

**Comment by sethmlarson** [comment]
Thanks for the review @zooba! I applied all of your suggestions and hopefully we're now getting consistent cross-platform SBOM generation :)

> Once those are done, there are still a few files showing up for me that aren't already in the SBOM. I can't immediately see why - maybe they're just lingering in my checkout? Or possibly they're being omitted from the POSIX run because of case sensitivity?
> 
>     Modules/expat/expat_config.h
>     Modules/_hacl/README.md
>     Modules/_hacl/refresh.sh
>     Lib/ctypes/macholib/README.ctypes
>     Lib/ctypes/macholib/fetch_macholib
>     Lib/ctypes/macholib/fetch_macholib.bat

These files shouldn't be in the SBOM, we track them in `PACKAGE_TO_FILES` (specifically `exclude=[...]` for each entry) and they were showing up in your SBOM because of path normalization. I believe I've fixed the issue by normalizing before checking the `exclude` lists.

**Comment by sethmlarson** [comment]
Thanks @zooba for testing this locally and finding the regex bug, I applied your patch.

**Comment by zooba** [comment]
Triggered an update mainly just to rerun tests and see if they pass. If not, we can look at whether to just ignore unrelated failures.

**Comment by sethmlarson** [comment]
@zooba Looks like things pass, ready to merge when you are :) Then I'll update the devguide PR accordingly.

**Comment by sethmlarson** [comment]
Thanks for the merge @zooba, I've created the backport for 3.12: https://github.com/python/cpython/pull/118435

**Pull request merged**